#01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

#02. Importing Data

In [2]:
#02 Importing Products CSV
df_prods = pd.read_csv(r'/Users/andrewruhnow/Documents/12-2023 Instacart Basket Analysis/02 Data/Original Data/products.csv', index_col = False)

In [3]:
#03 Importing Orders_Wrangled CSV
df_ords = pd.read_csv(r'/Users/andrewruhnow/Documents/12-2023 Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv', index_col = False)

In [4]:
# Create a dataframe
df_test = pd.DataFrame()

In [5]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [7]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [8]:
df_test['mix'] = df_test['mix'].astype('str')

In [9]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
df_prods.shape

(49693, 5)

In [13]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
# Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

4.5 Task Work

Q2:  Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [20]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Q2 Answer:  First column is "Unnamed: 0" not sure what the data in this column is for. Looking at the days_since_prior_order column the MIN is 0 and MAX is 3 but the 50% is 7 days which is higher than the MAX of 3 days.  Same for the order_number column the MIN and MAX is 1.0 but the percentile ranges are inconsistent 25% = 5.0 / 50% = 1.0 / 75% = 2.3.  

Q3:  Check for mixed-type data in your df_ords dataframe.

In [21]:
# Check for Mixed Data
print(df_ords)

         Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
1                         3                  7                    15.0  


In [22]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [23]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Q3 Answer:  No mixed-type data found in df_ords dataframe.

Q4:  If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

Q4 Answer:  No mixed-type data found in df_ords datafram.

Q5:  Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [25]:
# Missing Values Check in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [26]:
# creating new dataframe to check for NAN values in column days_since_prior_order
df_ords_prior_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [27]:
# view new dataframe
df_ords_prior_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

In [28]:
# verify that NAN values in days_since_prior_order column are connected to order_number column
df_ords_prior_nan['order_number'].describe()

count    206209.0
mean          1.0
std           0.0
min           1.0
25%           1.0
50%           1.0
75%           1.0
max           1.0
Name: order_number, dtype: float64

Q5 Answer: days_since_prior_order has 206,209 missing values and column is NaN when order_number equals 1.  This makes sense because there wouldn't be prior orders when it's a customers first and or only order.

Q6: Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [35]:
# Check number of rows in df_ords
df_ords.shape

(3421083, 8)

In [41]:
# Creating new column to separate out missing values
df_ords['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [43]:
df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                         2                  8                     NaN   
1                         3                  7                    15.0   
2                         3                 12                    21.0   
3                         4                  7                    29.0   
4                         4                 15                    28.0   
...                     ...                ...                     ...   
3421078                   5                 18                    29.0   
3421079                   4                 10                    30.0   
3421080                   1                 12                    18.0   
3421081                   1                 12                     7.0   
3421082                   6                 14                    30.0   

         first_order  
0               True  
1              False  
2              False  
3              False  
4              False  
...              ...  
3421078        False  
3421079        False  
3421080        False  
3421081        False  
3421082        False  

[3421083 rows x 9 columns]

Q6 Answer:  Didn't really need to address these missing values since we already know if order_number = 1 that days-since_prior_order would be NaN since there wouldn't be any prior orders.  I created an additional first_order column for better readability.    

Q7: Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [44]:
# Check for duplicates in df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [45]:
# View duplicates in df_ords
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

Q7 Answer:  There are no duplicates in df_ords dataframe

Q8: Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice.

Q8 Answer:  No duplicates in df_ords to address

Q9: Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [49]:
path = r'/Users/andrewruhnow/Documents/12-2023 Instacart Basket Analysis'

In [50]:
# Export df_ords as orders_checked_1
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked_1.csv'))

In [51]:
# Export df_prods as products_checked_1
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked_1.csv'))